In [6]:
import pandas as pd
from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()

X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

#Display the first five rows of the dataset
print(X.head())

#Print the feature names and check for missing values
print(X.columns)
X.isnull().sum()

#Generate summary statistics (mean, min, max, etc.)
X.describe()

   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25  
Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup',
       'Latitude', 'Longitude'],
      dtype='object')


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000


In [ ]:
#Split the dataset into training and test sets (80% training, 20% testing)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Train a linear regression model on the unscaled data
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

#Make predictions on the test set
y_pred = lin_reg.predict(X_test)
y_pred

array([0.71912284, 1.76401657, 2.70965883, ..., 4.46877017, 1.18751119,
       2.00940251], shape=(4128,))

In [8]:
#Evaluate model performance using the following metrics: MSE, RMSE, and R² Score
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score

mse = mean_squared_error(y_test, y_pred)
rmse = root_mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Unscaled Data Model:")
print(f"Mean Squared Error: {mse:.2f}")   #note to self: remember the 2f means to round to 2 decimal places
print(f"Root Squared Error: {rmse:.2f}")
print(f"R² Score: {r2:.2f}")


Unscaled Data Model:
Mean Squared Error: 0.56
Root Squared Error: 0.75
R² Score: 0.58


In [9]:
#Look at model coefficients to understand feature importance
print("Model Coefficients (Unscaled):")
print(pd.Series(lin_reg.coef_, 
                index=X.columns))

print("\nModel Intercept (Unscaled):")
print(lin_reg.intercept_)

Model Coefficients (Unscaled):
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64

Model Intercept (Unscaled):
-37.02327770606412


Interpretation Questions:

What does the R² score tell us about model performance?

It tells us how much of the variation in median house values is explained by the model. In this case, the R² value is about 0.58, which means the model explains roughly 58% of the variability in housing prices. This tells us that the model does a moderately good job capturing the relationship between the features and the target variable, but there is still about 42% of the variation that is not explained. This suggests that while the model is useful, there are other factors influencing housing prices that are not included in the model.

Which features seem to have the strongest impact on predictions based on the model’s coefficients?

Based on the magnitude of the coefficients, AveBedrms has the strongest positive impact on predicted house values (coefficient of 0.78), followed by MedInc. This suggests that areas with more average bedrooms and higher median income tend to have higher predicted house prices.

Latitude and Longitude also have relatively large coefficients (both negative), which indicates that geographic location plays a significant role in determining housing prices. This makes sense because housing values vary greatly depending on where the property is located in California.

On the other hand, Population, HouseAge, and AveOccup have very small coefficients, meaning they have a weaker impact on the model’s predictions compared to the other features.

How well do the predicted values match the actual values?

The predicted values match the actual values moderately well but not perfectly. The R² score of 0.58 lets us know that the model explains about 58% of the variability in housing prices. This means the model captures more than half of the overall trend in the data, but still a substantial portion of variation is still unexplained. 

The RMSE of 0.75 suggests that, on average, the model’s predictions differ from the actual values by about 0.75 units (which corresponds to roughly $75,000 since the target variable is measured in hundreds of thousands of dollars). This shows that while the model performs reasonably well for a simple linear regression, it still has noticeable prediction error.

In [ ]:
#Select three features from the dataset to build a simplified model
selected_features = ["MedInc", "AveBedrms", "Longitude"]

X_selected = X[selected_features]

#Train a new linear regression model using only these three features
X_train_s, X_test_s, y_train_s, y_test_s = train_test_split(X_selected, y, test_size=0.2, random_state=42)

simplified_linreg = LinearRegression()
simplified_linreg.fit(X_train_s, y_train_s)

y_pred_s = simplified_linreg.predict(X_test_s)

#Evaluate the performance of this simplified model and compare it to the full model
mse_s = mean_squared_error(y_test_s, y_pred_s)
rmse_s = root_mean_squared_error(y_test_s, y_pred_s)
r2_s = r2_score(y_test_s, y_pred_s)

print("Simplified Model:")
print(f"MSE: {mse_s:.2f}")
print(f"RMSE: {rmse_s:.2f}")
print(f"R²: {r2_s:.2f}")

Simplified Model:
MSE: 0.71
RMSE: 0.84
R²: 0.46


Explain your choice:
I selected AveBedrms, MedInc, and Longitude because they had some of the largest coefficients in the full model. AveBedrms had the largest coefficient, indicating it has a strong influence on predicted house values. MedInc also had a relatively large positive effect, which makes sense since higher-income areas tend to have more expensive homes. Longitude was included because its coefficient showed that geographic location plays a significant role in housing prices. These features were chosen based on both their statistical impact and real-world relevance.

Interpretation questions:

How does the simplified model compare to the full model?

The simplified model performs worse than the full model. Its R² score decreased from 0.58 to 0.46, meaning it explains less of the variation in housing prices. In addition, the RMSE increased from 0.75 to 0.84, which indicates larger prediction errors on average. This shows that removing features reduced the model’s predictive accuracy.

Would you use this simplified model in practice? Why or why not?

I would probably not use the simplified model if prediction accuracy is the main goal, since the full model performs better. However, the simplified model may still be useful if interpretability or simplicity is more important. Using fewer features makes the model easier to understand and implement, but it comes at the cost of reduced performance, especially taking into account that the reduction in the R² score was considerable in this case. 